### ISSUE TITLE
Why do PyTorch not support two successive inplace ReLU operations?

### ISSUE THREAD

**Issue by songyuc** [pytorch/pytorch#34429]
Hi guys,
I might a problem today, as I find that maybe PyTorch do not support two successive inplace ReLU operations, and here is an sample as,
```python
conv1 = nn.Conv2d(3, 128, kernel_size=1, bias=False)
bn = nn.BatchNorm2d(128)
relu1 = nn.ReLU(inplace = True)
relu2 = nn.ReLU(inplace = True)

input = torch.randn(10,3,8,8)
x = conv1(input)
x = bn(x)
x = relu1(x)

# the second successive inplace ReLU operation
x = relu2(x)

x = x.flatten()
x = torch.sum(x) 
x.backward()
```
I am wondering whether it is a bug or not?

Any answer or idea will be appreciated!

**Comment by vadimkantorov** [comment]
It's because of current version tracking limitations: https://discuss.pytorch.org/t/relu-dropout-inplace/13467

**Comment by songyuc** [comment]
Hi, @vadimkantorov Vadim!
Although two successive inplace ReLU operations are not supported, but inplace ReLU plus `+=` are supported as [https://github.com/pytorch/pytorch/issues/5687#issuecomment-412681482](https://github.com/pytorch/pytorch/issues/5687#issuecomment-412681482).
And I think these two kinds of operations have the same implementation strategy.

**Comment by vadimkantorov** [comment]
In the comment you cited, @colesbury explains the difference between += and two ReLU, so that may be the remaining piece of the puzzle. If you want to sidestep version tracking, you could use `.data`, but then it's up to you to make sure that computation is correct

**Comment by songyuc** [comment]
Thanks very much for @vadimkantorov your answer!

**Comment by albanD** [comment]
Two ReLU in a row or just one are the same thing. Why do you want to put two in a row?

Also as Vadim mentionned, the problem is that the ReLU needs the result to compute the backward so you should not override it.

I **strongly** advise against the use of `.data` because it has many many nasty side effects that you don't want (including breaking gradient computations).

**Comment by vadimkantorov** [comment]
> Why do you want to put two in a row?

@albanD I guess in practice it can happen as an artefact of some model decomposition choices, e.g. one module ends with ConvBNRelu, and the next module does ReLU+CosineSimilarity. 

Hope we get some version tracking sidestepping ways when `.data` is phased out! :)

**Comment by albanD** [comment]
We need to design and implement it. But it will be added before .data is removed for sure as we use .data for that purpose internally as well.

Btw in this case `.data` is not good as it will break the graph and prevent gradients from flowing ;)

**Comment by songyuc** [comment]
Hi, @albanD ,
I think @vadimkantorov 's argument is right, as I was porting an model in PaddlePaddle, where there are two successive inplace ReLUs in the code.
And the point is, now that PaddlePaddle supports this kind of operations, I do not think PyTorch is not responsible for this.
Besides, as one inplace ReLU function is able to undertake the task of gradient backward, there is no doubt that two successive inplace ReLUs could complete the same task, if the intermediate parameters were saved properly.

**Comment by albanD** [comment]
The thing is that the spirit of using inplace operation is to reduce memory. So in that same spirit, when we "save" variables, we don't copy them, we just take a reference to them.
If you later modify that tensor inplace, you will see the error you see now and you have to use the out-of-place version of your function.
Basically the idea is that we never do full copies of your Tensor under the hood, the user is responsible for doing so.
